src/core/ext/filters/client_channel/channel_connectivity.cc

```diff
@@ -152,15 +158,23 @@ static void partly_done(state_watcher* w, bool due_to_completion,
         w->error = error;
       }
       w->phase = CALLING_BACK_AND_FINISHED;
-      grpc_cq_end_op(w->cq, w->tag, w->error, finished_completion, w,
-                     &w->completion_storage);
+      end_op = true;
+      end_op_cq = w->cq;
```

Comment:
I don't think the rest of these variables are needed.  We can still access these fields directly from the `w` struct when we get down to line 174.

Reply:
The idea was to capture the whole status before leaving the lock, so making a snapshot of all of the dependencies that are later used by the `grpc_cq_end_op` function.

Indeed was a suggestion from @vjpai

Reply:
We shouldn't access them down there as we've let the lock go by then. The mu on this struct is currently being used to protect the struct and we can't hold the mu when doing the end_op so we can't be accessing the struct either.

Reply:
This is a common construct used in our code and many others when we have to make a decision while a struct is protected but not execute on that decision until the struct is no longer protected.